agent/src/softwarecontaineragent.cpp

```diff
@@ -193,88 +196,145 @@ bool SoftwareContainerAgent::checkJob(pid_t pid, CommandJob *&result)
             return true;
         }
     }
+
     log_warning() << "Unknown PID: " << pid;
     return false;
 }
 
-void SoftwareContainerAgent::writeToStdIn(pid_t pid, const std::vector<uint8_t> &bytes)
+bool SoftwareContainerAgent::writeToStdIn(pid_t pid, const std::vector<uint8_t> &bytes)
 {
     CommandJob *job = nullptr;
-    if (checkJob(pid, job)) {
-        log_debug() << "writing bytes to process with PID:" << job->pid() << " : " << bytes;
-        write(job->stdin(), bytes.data(), bytes.size());
+    if (!checkJob(pid, job)) {
+        log_error() << "Unknown pid " << pid;
+        return false;
     }
+
+    log_debug() << "writing bytes to process with PID:" << job->pid() << " : " << bytes;
+    write(job->stdin(), bytes.data(), bytes.size());
+    return true;
 }
 
 pid_t SoftwareContainerAgent::launchCommand(ContainerID containerID, uid_t userID, const std::string &cmdLine, const std::string &workingDirectory, const std::string &outputFile, const EnvironmentVariables &env, std::function<void (pid_t, int)> listener)
 {
     profilefunction("launchCommandFunction");
-    SoftwareContainer *container;
-    if (checkContainer(containerID, container)) {
-        auto job = new CommandJob(*container, cmdLine);
-        // Capturing this leaves an open pipe for every container, even
-        // after it has terminated.
-        // job->captureStdin();
-        job->setOutputFile(outputFile);
-        job->setUserID(userID);
-        job->setEnvironmentVariables(env);
-        job->setWorkingDirectory(workingDirectory);
-        job->start();
-        addProcessListener(m_connections, job->pid(), [listener](pid_t pid, int exitCode) {
-            listener(pid, exitCode);
-        }, m_mainLoopContext);
-
-        m_jobs.push_back(job);
-
-        profilepoint("launchCommandEnd");
-
-        return job->pid();
+    SoftwareContainer *container = nullptr;
+    if (!checkContainer(containerID, container)) {
+        log_error() << "Invalid container ID " << containerID;
+        return INVALID_PID;
+    }
+
+    // Set up a CommandJob for this run in the container
+    auto job = new CommandJob(*container, cmdLine);
+    // Capturing this leaves an open pipe for every container, even
+    // after it has terminated.
+    // job->captureStdin();
```

Comment:
As said in the comment above, having it uncommented leaks an open pipe. We have a story for fixing this.